server/profiler.py

```diff
@@ -0,0 +1,93 @@
+import asyncio
+import cProfile
+import logging
+
+from server.config import config
+from server.decorators import with_logger
+from server.timing import at_interval
+
+
+@with_logger
+class Profiler:
+    def __init__(
+        self, interval, player_service, duration=2, max_count=300, outfile="profile.txt"
+    ):
+        self.profiler = cProfile.Profile()
+        self.interval = interval
+        self.duration = duration
+        self.current_count = 0
+        self.max_count = max_count
+
+        self._player_service = player_service
+        self._outfile = outfile
+
+        self._running = False
+        self._task = None
+        self._loop = asyncio.get_running_loop()
+
+    def start(self):
+        self._running = True
+        if self._task is None:
+            self._task = asyncio.create_task(Profiler._next_run(self))
+
+    async def _next_run(self):
+        await asyncio.sleep(self.interval)
+
+        if not self._running:
+            return
+        self.current_count += 1
+        await self._run()
+
+        if self.current_count < self.max_count and self._running:
+            self._task = asyncio.create_task(Profiler._next_run(self))
+
+    async def _run(self):
+        if len(self._player_service) > 1000:
+            self._logger.info(
+                "Refusing to profile under high load %i/%i",
+                self.current_count,
+                self.max_count,
+            )
+            return
+
+        self._logger.info("Starting profiler")
+        self.profiler.enable()
+        await asyncio.sleep(self.duration)
+        self.profiler.disable()
+
+        self._logger.info("Done profiling %i/%i", self.current_count, self.max_count)
+        if self._outfile is not None:
+            self.profiler.dump_stats(self._outfile)
```

Comment:
I left that to garbage collection (since the instance of the wrapping `Profiler` class isn't referenced anywhere and only keeps scheduling itself it should go out of scope when it's done), but if it's that large it's maybe a good idea to do explicitly...